cli/internal/oauth/api/api.go

```diff
@@ -0,0 +1,229 @@
+)
+
+type OAuthAPI interface {
+	GetDeviceCode(ctx context.Context, audience string) (State, error)
+	WaitForDeviceToken(ctx context.Context, state State) (TokenResponse, error)
+	RevokeToken(ctx context.Context, refreshToken string) error
+	GetAutoPAT(ctx context.Context, audience string, res TokenResponse) (string, error)
+}
+
+// API represents API interactions with Auth0.
+type API struct {
+	// TenantURL is the base used for each request to Auth0.
+	TenantURL string
+	// ClientID is the client ID for the application to auth with the tenant.
+	ClientID string
+	// Scopes are the scopes that are requested during the device auth flow.
+	Scopes []string
+}
+
+// TokenResponse represents the response of the /oauth/token route.
+type TokenResponse struct {
+	AccessToken      string  `json:"access_token"`
+	IDToken          string  `json:"id_token"`
+	RefreshToken     string  `json:"refresh_token"`
+	Scope            string  `json:"scope"`
+	ExpiresIn        int     `json:"expires_in"`
+	TokenType        string  `json:"token_type"`
+	Error            *string `json:"error,omitempty"`
+	ErrorDescription string  `json:"error_description,omitempty"`
+}
+
+var ErrTimeout = errors.New("timed out waiting for device token")
+
+// GetDeviceCode initiates the device-code auth flow with the tenant.
+// The state returned contains the device code that the user must use to
+// authenticate, as well as the URL to visit, etc.
+func (a API) GetDeviceCode(ctx context.Context, audience string) (State, error) {
+	data := url.Values{
+		"client_id": {a.ClientID},
+		"audience":  {audience},
+		"scope":     {strings.Join(a.Scopes, " ")},
+	}
+
+	deviceCodeURL := a.TenantURL + "/oauth/device/code"
+	resp, err := postForm(ctx, deviceCodeURL, strings.NewReader(data.Encode()))
+	if err != nil {
+		return State{}, err
+	}
+	defer func() {
+		_ = resp.Body.Close()
+	}()
+
+	if resp.StatusCode != http.StatusOK {
+		return State{}, tryDecodeOAuthError(resp)
+	}
+
+	var state State
+	err = json.NewDecoder(resp.Body).Decode(&state)
+	if err != nil {
+		return state, fmt.Errorf("failed to get device code: %w", err)
+	}
+
+	return state, nil
+}
+
+func tryDecodeOAuthError(resp *http.Response) error {
+	var body map[string]any
+	if err := json.NewDecoder(resp.Body).Decode(&body); err == nil {
+		if errorDescription, ok := body["error_description"].(string); ok {
+			return errors.New(errorDescription)
+		}
+	}
+	return errors.New("unexpected response from tenant: " + resp.Status)
+}
+
+// WaitForDeviceToken polls the tenant to get access/refresh tokens for the user.
+// This should be called after GetDeviceCode, and will block until the user has
+// authenticated or we have reached the time limit for authenticating (based on
+// the response from GetDeviceCode).
+func (a API) WaitForDeviceToken(ctx context.Context, state State) (TokenResponse, error) {
+	ticker := time.NewTicker(state.IntervalDuration())
+	timeout := time.After(time.Duration(state.ExpiresIn) * time.Second)
```

Comment:
Since we already respect context cancellation there - how about just using a `context.WithTimeout`, possibly on the caller side?

Reply:
I like that idea (but haven't dug in too deep on that matter)

Reply:
Hmm, I generally prefer this oauth-tenant timer logic to be in one place rather than spread out. We take the `State` here and set up timers based on that. This will always be the same (regardless of caller) since this is coming from the tenant, so I don't know why we'd do that.

Also, I think it's a different case to have "context was cancelled" here rather than "it took longer than the tenant said it could", so I'd rather not abstract that away from this bit of code.

Reply:
Fine by me